lib/drivers/abbott/freeStyleLibreData.js

```diff
@@ -0,0 +1,466 @@
+
+  handleCompressedDatabase(aapPacket) {
+    let decompressedBuffer = new Buffer(1);
+    let compressedOffset = 0;
+
+    // copy table ID
+    decompressedBuffer[0] = aapPacket.data[compressedOffset];
+    compressedOffset++;
+
+    while (compressedOffset < aapPacket.dataLength) {
+      const blockType = aapPacket.data[compressedOffset];
+      compressedOffset++;
+
+      // parse 24 bit little endian block length
+      let blockLength = aapPacket.data[compressedOffset]
+        | (aapPacket.data[compressedOffset + 1] << 8)
+        | (aapPacket.data[compressedOffset + 2] << 16);
+      compressedOffset += 3;
+
+      blockLength *= 4; // convert number of uint32 values to number of uint8 values
+
+      if (blockType === COMPRESSION_TYPE.UNCOMPRESSED) {
+
+        decompressedBuffer = Buffer.concat(
+          [decompressedBuffer, aapPacket.data.slice(compressedOffset, compressedOffset + blockLength)]);
+        compressedOffset += blockLength;
+
+      } else if (blockType === COMPRESSION_TYPE.ZERO_COMPRESSED) {
+
+        decompressedBuffer = Buffer.concat([decompressedBuffer, Buffer.alloc(blockLength)]);
+        compressedOffset += blockLength;
+
+      } else {
+        debug('handleCompressedDatabase: failed to decompress!');
+        return;
+      }
+    }
+
+    // build decompressed AAP packet to process
+    const decompressedAapPacket = {
+      packetLength: aapPacket.packetLength - aapPacket.dataLength + aapPacket.data.length,
+      data: decompressedBuffer,
+      dataLength: aapPacket.data.length,
+      opCode: OP_CODE.GET_DATABASE
+    };
+
+    this.handleDatabase(decompressedAapPacket);
+  }
+
+  handleDatabase(aapPacket) {
+    if (aapPacket.dataLength === 0) {
+      return;
+    }
+
+    let offset = 0;
+    const databaseTableId = aapPacket.data[offset];
+    offset += 1;
+
+    const headerFields = struct.unpack(aapPacket.data, offset, FORMAT.RECORD_HEADER,
+      ['recordNumber', 'recordType', 'isTimeValid', 'readerTime', 'userTimeOffset']);
+    headerFields.isTimeValid = ((headerFields.isTimeValid & 0x80) > 0);
+    offset += FORMAT_LENGTH.RECORD_HEADER;
+
+    // calculate actual 32bit record number from 16bit header record number and next wrap around number
+    headerFields.recordNumber = this.dbRecordNumberNextWrap[databaseTableId] - (0x10000 - headerFields.recordNumber);
+
+    // find the lowest record number in the results database
+    if (databaseTableId === DB_TABLE_ID.GLUCOSE_RESULT) {
+      this.numResultRecords++;
+      this.oldestResultRecordNumber = Math.min(this.oldestResultRecordNumber, headerFields.recordNumber);
+    }
+
+    const dateTime = this.getDateTime(headerFields.readerTime, headerFields.userTimeOffset);
+
+    if (headerFields.recordType === DB_RECORD_TYPE.TIME_CHANGE_RESULT) {
+      const timeChangeFields = struct.unpack(aapPacket.data, offset, FORMAT.TIME_CHANGE,
+        ['oldReaderTime', 'oldUserTimeOffset', 'valid', 'unused', 'CRC16']);
+
+      // TODO: validate CRC16
```

Comment:
Do you have any example data sets and their corresponding CRCs? I think it's worth doing the extra validation if it's not too hard to figure out. The Medtronic drivers similarly have a one byte checksum, and several layers of CRC8 or CCITT checksums, all of which are validated.